packages/engine/src/avatar/components/TransparencyDitheringComponent.tsx

```diff
@@ -0,0 +1,165 @@
+/*
+CPAL-1.0 License
+
+The contents of this file are subject to the Common Public Attribution License
+Version 1.0. (the "License"); you may not use this file except in compliance
+with the License. You may obtain a copy of the License at
+https://github.com/EtherealEngine/etherealengine/blob/dev/LICENSE.
+The License is based on the Mozilla Public License Version 1.1, but Sections 14
+and 15 have been added to cover use of software over a computer network and 
+provide for limited attribution for the Original Developer. In addition, 
+Exhibit A has been modified to be consistent with Exhibit B.
+
+Software distributed under the License is distributed on an "AS IS" basis,
+WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License for the
+specific language governing rights and limitations under the License.
+
+The Original Code is Ethereal Engine.
+
+The Original Developer is the Initial Developer. The Initial Developer of the
+Original Code is the Ethereal Engine team.
+
+All portions of the code written by the Ethereal Engine team are Copyright © 2021-2023 
+Ethereal Engine. All Rights Reserved.
+*/
+
+import { Entity, defineComponent, getOptionalComponent, useComponent, useEntityContext } from '@etherealengine/ecs'
+import { defineState, getMutableState, matches, useHookstate } from '@etherealengine/hyperflux'
+import { matchesVector3 } from '@etherealengine/spatial/src/common/functions/MatchesUtils'
+import { addOBCPlugin } from '@etherealengine/spatial/src/common/functions/OnBeforeCompilePlugin'
+import { RendererState } from '@etherealengine/spatial/src/renderer/RendererState'
+import { MeshComponent } from '@etherealengine/spatial/src/renderer/components/MeshComponent'
+import { isArray } from 'lodash'
+import React, { useEffect } from 'react'
+import { FrontSide, Material, Vector3 } from 'three'
+import { SceneComponent } from '../../scene/components/SceneComponent'
+import { useModelSceneID } from '../../scene/functions/loaders/ModelFunctions'
+import {
+  ditheringAlphatestChunk,
+  ditheringFragUniform,
+  ditheringVertex,
+  ditheringVertexUniform
+} from '../functions/ditherShaderChunk'
+
+export enum ditherCalculationType {
+  worldTransformed = 1,
+  localPosition = 0
+}
+
+export const transparencyDitheringState = defineState({
+  name: 'transparencyDitheringState',
+  initial: { materialIds: {} as Record<Entity, string[]> }
+})
+
+export const maxDitherPoints = 4 //should be equal to the length of the vec3 array in the shader
+export const TransparencyDitheringComponent = Array.from({ length: maxDitherPoints }, (_, i) => {
```

Comment:
do we need multiple components here, or do we just want to make the data an array?

Reply:
I went with an array of components as opposed to an array of component values because it seemed straightforward and easier to introspect than either of the other options I considered in that vein. When you have arrays of uniforms in the component you can expect the arrays to end up with different lengths, which would likely require helper functions for adding/removing dithering points to avoid some confusing inconsistency. I saw using set/remove component for this as a UX simplifcation that ensures there will always be values that correspond to a dither point for a given index, without abstracting the data into an array of objects within the component. Is using an array of components in this context bad practice?

Reply:
no worries, it shouldn't be much of a performance concern.